### DISCUSSION TITLE
GStreamer Problem related to pkgconfig files?

### DISCUSSION THREAD

**Discussion by KANIYONIKA** [Homebrew/discussions#3740] [Getting started]
I was developing an app using GStreamer and ran into this issue.

https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/1159

I checked the contents of this issue and found that the issue needs to be resolved on the homebrew side.
> yes. I guess we should've blocked 1.20.1 on -Wl,-rpath,<path> being added to the pkgconfig files. We should do it now, before 1.20.2 (as a blocker).

I am not sure how to go about getting this issue resolved.
I apologize if I am rude. Please advise me.

**Comment by carlocab** [comment]
It seems like backporting the patch referenced in the issue might help: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/commit/97655eb8ac70b6de0b672e8df73ece5c901371d3

Try to see if rebuilding `gstreamer` with that patch helps. If it does we can apply it to the formula and fix it for everyone else using it.

**Reply by KANIYONIKA** [reply]
Thanks for the reply.

Um...it's difficult for me. It seems to take a lot of time.
I will think about it a lot.

Anyway, thanks.

**Reply by KANIYONIKA** [reply]
I checked the latest "meson.build" and the following statement remains

```
pkgconfig_libs = []
if host_system == 'darwin'
  pkgconfig_libs = ['-Wl,-rpath,${libdir}']
endif
```

This may be out of line to ask you, but if you can please advise me.

Is it possible that this problem could be solved by eliminating these lines and doing a build?

Also, if so, as you say, I would like to test if that would work. Please let me know what documentation would be helpful in testing. I am ashamed to say that I don't know how to do a build in homebrew using `.build`.

Thanks in advance.

"meson.build".
https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gstreamer/meson.build

**Reply by nirbheek** [reply]
@carlocab that fix will not be sufficient, because it will only fix loading of `libgstreamer-1.0.0.dylib` (and other libraries that are in the same prefix as that). The actual problem is that the brew formula puts each gstreamer module into separate prefixes: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/1159#note_1343088 so you need to add `-Wl,-rpath,${libdir}` to every pkg-config file so that RPATH entries are added to every libdir prefix.

This split structure also breaks loading of plugins, since that is not a supported gstreamer configuration. We try to automatically detect the 'system path' for plugins and load everything in there, so we need all of gstreamer to be installed into a single prefix. 

First, we detect libdir by finding the location of `libgtreamer-1.0.dylib`: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gstreamer/gst/gstregistry.c#L1604-1623

Then we load plugins from `$libdir/gstreamer-1.0`: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gstreamer/gst/gstregistry.c#L1758-1768

All other plugin directories have to be explicitly pointed to by the user, which means that by default GStreamer in brew doesn't load any plugins and isn't usable. You've added a `gst-inspect-1.0` wrapper script that sets `GST_PLUGIN_SYSTEM_PATH` which just works around the issue and doesn't solve it. You should get rid of that.

So I recommend either sticking to the directory structure that gstreamer wants, or patching all this code to do what Homebrew wants. IMO the former is what you should do. There is no technical advantage to putting each module into a separate prefix.

I recommend adding some post-install tests:

1. Build and run a test gstreamer program such as [`gst-play.c`](https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gst-plugins-base/tools/gst-play.c) that links to gstreamer and gst-plugins-base.
2.  Perform a non-wrapped `gst-inspect-1.0` invocation to assert that plugins are actually getting loaded. You'll want to check that the output contains something more than `coreelements` at least.

**Reply by carlocab** [reply]
Hey @nirbheek, thanks for dropping by.

Putting each module in a separate prefix is just a consequence of how Homebrew is designed. The only way to put them all in the same prefix (without rewriting large parts of `brew`, that is) is if the gstreamer formula included all of its plugins. That would be really impractical -- updates to any of its components would require rebuilding everything. Also, users would need to install all the plugins even if they were only interested in one of them. Still, I guess it might be better than shipping something broken.

**Reply by nirbheek** [reply]
@carlocab ever since gstreamer [merged all the repos into one monorepo](https://gstreamer.freedesktop.org/documentation/frequently-asked-questions/mono-repository.html), you can actually build all of gstreamer at once in a single formula. You can build tags from git instead of using the split tarballs (which only exist so that distros don't have to change their packaging): https://gitlab.freedesktop.org/gstreamer/gstreamer

Even before this, updates (releases) to components of gstreamer always happened together.

The remaining problem will be libnice (which is a separate project) and the gstreamer plugin that it ships -- which is necessary for webrtc support (a very commonly-used feature in gstreamer). Maybe that formula can print instructions on what the user should do to get that plugin to be discoverable.

As an aside, I believe gstreamer webrtc support is broken in brew right now since it doesn't ship all the necessary plugins for that: sctp, srtp, dtls, nice.

We generally have to recommend that people not use brew because of all these papercuts. It would be nice to have Brew working as well as the official gstreamer macOS packages :slightly_smiling_face:

**Reply by carlocab** [reply]
Oh, that's great news. Thanks for the info. I take it this means `meson ... && ninja ...` from a Git tag should just Do The Right Thing then?

Additionally, from the formulae that use gstreamer:
```
❯ brew uses --eval-all --include-build --include-test --recursive gstreamer
aravis                        gst-devtools                  gst-plugins-base              gst-python                    pympress
clutter-gst                   gst-editing-services          gst-plugins-good              gst-rtsp-server               spice-gtk
efl                           gst-libav                     gst-plugins-rs                libnice                       virt-manager
gradio                        gst-plugins-bad               gst-plugins-ugly              pdfpc                         webkitgtk
```
I take it only the ones that are prefixed with `gst-` are the plugins that will be replaced by a monorepo build? Looks like gst-libav is going to pull in FFmpeg, which isn't ideal, but oh well.

Apologies for all the questions -- I don't actually use gstreamer, so I'm not familiar at all with it.

**Reply by nirbheek** [reply]
> I take it this means meson ... && ninja ... from a Git tag should just Do The Right Thing then?

Yes. You may like to use `--wrap-mode=nodownload` so that meson doesn't download anything behind your back (I guess you already have abstractions for this for use with other meson formula?), and you want to look at `meson_options.txt` for things to enable. You can pass `-Dbase=enabled` etc to enable each component.

Another thing distros do is to pass `--auto-features=enabled` and then manually disable everything you don't want. That way you don't have any 'automagic' dependencies. But that's _a lot_ of work, since there's dozens of plugins. You can skip that and just specifically enable the plugins that you want to ship, like so: `-Dbad=enabled -Dgst-plugins-bad:srtp=enabled`.

> I take it only the ones that are prefixed with `gst-` are the plugins that will be replaced by a monorepo build?

Yes, except for `gst-plugins-rs`, which is a separate repository. However, you *can* actually build it using this setup, by passing `-Drs=enabled`. Since you probably want to control what's downloaded yourself, you will want to set up `subprojects/gst-plugins-rs` yourself with, f.ex., `meson subprojects download gst-plugins-rs` before running `meson setup`. Starting from 1.22 (and the 1.21 dev releases), there will be a corresponding tag in gst-plugins-rs for each gstreamer release. In 1.20 you're supposed to use the 0.8 release branch.

> Apologies for all the questions -- I don't actually use gstreamer, so I'm not familiar at all with it.

That's ok, I'm happy someone is having a look at fixing up the gstreamer brew formulae :)

**Reply by KANIYONIKA** [reply]
Thank you for your discussion!!
I am ashamed to say that I don't think I can do it, so we will wait for someone else to do it.

**Reply by carlocab** [reply]
I'll try to get this done at https://github.com/Homebrew/homebrew-core/pull/113061.

**Reply by KANIYONIKA** [reply]
Thank you✨Nice Guy!!

**Comment by danburkert** [comment]
Setting `env GST_PLUGIN_PATH=/opt/homebrew/lib` appears to be a workaround, although it causes gstreamer to issue a warning log for every non-plugin dylib in that directory which it can't load.  Perhaps setting it to a more targeted list of the plugin dylibs would fix that.

**Reply by carlocab** [reply]
Try
```zsh
plugin_libdirs=("$(brew --prefix)"/opt/gst-*/lib)
export GST_PLUGIN_PATH="${(j|:|)plugin_libdirs}"
```
⚠️ Zsh syntax, may not work on other shells

**Comment by carlocab** [comment]
I've opened Homebrew/homebrew-core#125996 to try to get this done. @nirbheek, I'd appreciate it if you could have a look. In particular, there are options we used to pass to what will now be subprojects which I'm not sure how to handle from a `meson` invocation.

**Reply by carlocab** [reply]
Hmmm, ok, one issue is that `gst-plugins-bad` depends on `libnice`. Except that `libnice` also depends on `gstreamer`, so once `gst-plugins-bad` is built as part of the `gstreamer` monorepo, we have a circular dependency:
```
gstreamer -> libnice -> gstreamer
```
Not quite sure how to resolve this, so I guess I'm blocked here.

**Comment by carlocab** [comment]
I've merged Homebrew/homebrew-core#125996. Give it about ~20 minutes for the API to update then do `brew update && brew upgrade gstreamer`.

**Comment by mraerino** [comment]
i'm still seeing issue with using gstreamer in a Rust application:

```
% PKG_CONFIG_PATH=/opt/homebrew/lib/pkgconfig:/opt/homebrew/share/pkgconfig cargo build -v
-- snip --
     Running `rustc --crate-name gstreamer --edition=2021 /Users/marcus/.cargo/registry/src/index.crates.io-6f17d22bba15001f/gstreamer-0.20.5/src/lib.rs --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="serde"' --cfg 'feature="serde_bytes"' --cfg 'feature="v1_16"' --cfg 'feature="v1_18"' --cfg 'feature="v1_20"' --cfg 'feature="v1_22"' -C metadata=fa913bb4cf74ffa4 -C extra-filename=-fa913bb4cf74ffa4 --out-dir /Users/marcus/src/mraerino/merdia/target/debug/deps -L dependency=/Users/marcus/src/mraerino/merdia/target/debug/deps -L native=/opt/homebrew/Cellar/glib/2.76.2/lib -L native=/opt/homebrew/opt/gettext/lib -L native=/opt/homebrew/Cellar/glib/2.76.2/lib -L native=/opt/homebrew/opt/gettext/lib -L native=/opt/homebrew/Cellar/glib/2.76.2/lib -L native=/opt/homebrew/opt/gettext/lib -L native=/opt/homebrew/Cellar/glib/2.76.2/lib -L native=/opt/homebrew/opt/gettext/lib -L native=/opt/homebrew/Cellar/gstreamer/1.22.2/lib -L native=/opt/homebrew/Cellar/glib/2.76.2/lib -L native=/opt/homebrew/opt/gettext/lib`
-- snip --

% ./target/debug/mybin

(<unknown>:64557): GStreamer-WARNING **: 22:52:10.815: External plugin loader failed. This most likely means that the plugin loader helper binary was not found or could not be run. You might need to set the GST_PLUGIN_SCANNER environment variable if your setup is unusual. This should normally not be required though.
thread 'tokio-runtime-worker' panicked at 'Could not find webrtcbin', src/main.rs:129:50
```

in contrast, it works when using the libraries directly provided by gstreamer:

```
% PKG_CONFIG_PATH=/Library/Frameworks/GStreamer.framework/Versions/Current/lib/pkgconfig cargo build -v
     Running `rustc --crate-name gstreamer --edition=2021 /Users/marcus/.cargo/registry/src/index.crates.io-6f17d22bba15001f/gstreamer-0.20.5/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=122 --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="serde"' --cfg 'feature="serde_bytes"' --cfg 'feature="v1_16"' --cfg 'feature="v1_18"' --cfg 'feature="v1_20"' --cfg 'feature="v1_22"' -C metadata=fa913bb4cf74ffa4 -C extra-filename=-fa913bb4cf74ffa4 --out-dir /Users/marcus/src/mraerino/merdia/target/debug/deps -L dependency=/Users/marcus/src/mraerino/merdia/target/debug/deps --cap-lints allow -L native=/Library/Frameworks/GStreamer.framework/Versions/Current/lib/pkgconfig/../../lib -L native=/Library/Frameworks/GStreamer.framework/Versions/Current/lib/pkgconfig/../../lib -L native=/Library/Frameworks/GStreamer.framework/Versions/Current/lib/pkgconfig/../../lib -L native=/Library/Frameworks/GStreamer.framework/Versions/Current/lib/pkgconfig/../../lib -L native=/Library/Frameworks/GStreamer.framework/Versions/Current/lib/pkgconfig/../../lib`
```

is there anything i can do to debug this?

**Reply by carlocab** [reply]
What is the output of `brew config`, `brew doctor`, and `brew info gstreamer`?

**Reply by mraerino** [reply]
```
% brew config
HOMEBREW_VERSION: 4.0.16
ORIGIN: https://github.com/Homebrew/brew
HEAD: 01e0c20d01c6aab28d844280d130f0e745e2ddd9
Last commit: 4 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 3a4d62846a2f0a785cb375f2d6b1599697db755a
Core tap last commit: 3 months ago
Core tap branch: master
Core tap JSON: 04 May 21:50 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.40.1 => /opt/homebrew/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
Rosetta 2: false
```

```
% brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  pycparser
  cffi

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  wireshark

Warning: You have an unnecessary local Cask tap.
This can cause problems installing up-to-date casks.
Please remove it by running:
  brew untap homebrew/cask

Warning: You have an unnecessary local Core tap!
This can cause problems installing up-to-date formulae.
Please remove it by running:
 brew untap homebrew/core
```

```
% brew info gstreamer
==> gstreamer: stable 1.22.2 (bottled), HEAD
Development framework for multimedia applications
https://gstreamer.freedesktop.org/
/opt/homebrew/Cellar/gstreamer/1.22.2 (1,269 files, 168MB) *
  Poured from bottle using the formulae.brew.sh API on 2023-05-03 at 20:42:19
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/gstreamer.rb
License: LGPL-2.0-or-later and LGPL-2.1-or-later and MIT
==> Dependencies
Build: bison ✔, cargo-c ✘, gobject-introspection ✔, meson ✘, nasm ✘, ninja ✘, pkg-config ✔, rust ✘, yasm ✘
Required: cairo ✔, dav1d ✔, faac ✔, faad2 ✔, fdk-aac ✔, ffmpeg ✔, flac ✔, gettext ✔, glib ✔, glib-networking ✔, graphene ✔, gtk+3 ✔, gtk4 ✔, jpeg-turbo ✔, json-glib ✔, lame ✔, libogg ✔, libpng ✔, libpthread-stubs ✔, libshout ✔, libsoup ✔, libusrsctp ✔, libvorbis ✔, libvpx ✔, openssl@1.1 ✔, opus ✔, orc ✔, pango ✔, pygobject3 ✔, python@3.11 ✔, rav1e ✔, rtmpdump ✔, speex ✔, srtp ✔, taglib ✔, theora ✔, x264 ✔, xz ✔, musepack ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
All gst-* GStreamer plugins are now bundled in this formula.
For GStreamer to find your own plugins, add their paths to `GST_PLUGIN_PATH`.
For example, if you have plugins in `~/.local/lib/gstreamer-1.0`:
  export GST_PLUGIN_PATH="~/.local/lib/gstreamer-1.0"

Do not install plugins into GStreamer's prefix. They will be deleted
by `brew upgrade`.
==> Analytics
install: 111 (30 days), 7,504 (90 days), 86,762 (365 days)
install-on-request: 70 (30 days), 4,656 (90 days), 53,536 (365 days)
build-error: 0 (30 days)
```

**Reply by carlocab** [reply]
Ah yes, I see, this is the problem:
```
'Could not find webrtcbin'
```
We don't enable webrtc at the moment because of a circular dependency issue.

There's some discussion [here](https://github.com/Homebrew/homebrew-core/pull/125996#discussion_r1180019113) about a fix, but I'm waiting to hear back from @nirbheek there.

**Reply by mraerino** [reply]
can i open a tracking issue for that?

**Reply by carlocab** [reply]
You can, but I'm not very sure how one is going to help.

**Reply by razvanphp** [reply]
Any news about the webrtc plugin / libnice support?

**Reply by nirbheek** [reply]
@carlocab I've opened a merge request with a fix for this issue: https://gitlab.freedesktop.org/libnice/libnice/-/merge_requests/271

I recommend that you split libnice into two packages:

1. libnice
    - Build with `-Dgstreamer=disabled`
2. libnice-gstreamer
    - Apply the changes from [the MR](https://gitlab.freedesktop.org/libnice/libnice/-/merge_requests/271)
    - Build with `-Dgstreamer=enabled -Dgstreamer-plugin-only=true`

Don't wait for the patch to get upstreamed, I will help you maintain it.

**Reply by nirbheek** [reply]
@carlocab poke again  :slightly_smiling_face: or could you please point this to the current formula maintainer for gstreamer? We regularly get people on our support channels asking about this :sweat_smile:

**Reply by carlocab** [reply]
All formulae in Homebrew/core are community maintained, so each formula doesn't have a specific maintainer. But I've opened https://github.com/Homebrew/homebrew-core/pull/171902 to see if we can do this.

**Reply by nirbheek** [reply]
Thanks Carlo!

**Comment by jessbowers** [comment]
https://github.com/Homebrew/homebrew-core/pull/171902 is merged - is this fixed now?

**Reply by carlocab** [reply]
Probably -- haven't heard any more issues about it since then. @nirbheek?

**Reply by nirbheek** [reply]
This was fixed with https://gitlab.freedesktop.org/gstreamer/cerbero/-/merge_requests/1485, which went into GStreamer 1.24.5

**Reply by nirbheek** [reply]
Ah, that was an early-morning comment. That fix was for cerbero, which is unrelated to homebrew (ofc). I don't know if it's fixed in brew, but it must be fine now if the DYLIB ID is not `@rpath/libfoo.dylib`.

**Reply by lu-zero** [reply]
As today is possible to build and run the iced_player example w/out problems.
